pyanaconda/modules/subscription/subscription.py

```diff
@@ -287,6 +296,68 @@ def set_system_purpose_data(self, system_purpose_data):
+        and the target system, as the token transfer installation task will
+        make sure any system purpose configuration file created in the installation
+        environment will be transferred to the target system.
+
+        We also need to avoid running system purpose configuration again after
+        a successful subscription attempt, as subscription can actually change
+        the system purpose attached to the system via system purpose values
+        attached to an activation key. If we re-run the system purpose task
+        on the installed system, we would basically overwrite these changes.
+        """
+        return self._is_system_purpose_applied
+
+    def set_is_system_purpose_applied(self, system_purpose_applied):
+        """Set if system purpose is applied.
+
+        :param bool system_purpose_applied: True if applied, False otherwise
+
+        NOTE: We keep this as a private method, called by the completed signal of the
+              task that applies system purpose information on the system.
+        """
+        self._is_system_purpose_applied = system_purpose_applied
+        self.is_system_purpose_applied_changed.emit()
+        # as there is no public setter in the DBus API, we need to emit
+        # the properties changed signal here manually
+        self.module_properties_changed.emit()
+        log.debug("System purpose is applied set to: %s", system_purpose_applied)
+
+    def _apply_syspurpose(self):
+        """Apply system purpose information to the installation environment.
+
+        If this method is called, then the token transfer installation task will
+        make sure to transfer the result, so the system purpose installation task
+        does not have to run afterwards.
+        For this reason we record if this method has run via the
+        set_is_system_purpose_applied() method.
+        """
+        log.debug("subscription: Applying system purpose data")
+        task = SystemPurposeConfigurationTask(sysroot="/",
+                                              system_purpose_data=self.system_purpose_data)
+        if task.run():
+            # set system purpose as applied once the task finishes running
+            self.set_is_system_purpose_applied(True)
+        else:
+            log.error("subscription: Failed to apply system purpose data")
+
+    def set_system_purpose_with_task(self):
+        """Set system purpose for the installed system with an installation task.
+
+        :param sysroot: a path to the root of the installed system
+        :return: a DBus path of an installation task
+        """
+        task = SystemPurposeConfigurationTask(sysroot=conf.target.system_root,
+                                              system_purpose_data=self.system_purpose_data)
+        # set system purpose as applied once the task successfully finishes running
+        task.succeeded_signal.connect(lambda: self.set_is_system_purpose_applied(True))
```

Comment:
I guess you want to set `self.set_is_system_purpose_applied(task.get_result())`.

Reply:
Good point, changed. :)